service_container/service_subscribers.rst

```diff
@@ -0,0 +1,403 @@
+coincide with the service locator::
+
+    // src/AppBundle/Updates/FooUpdater.php
+    namespace AppBundle\Updates;
+
+    class FooUpdater extends AbstractUpdater
+    {
+        public function update($entity)
+        {
+            // ...
+
+            $body = $this->render('Emails/foo_update.html.twig', [/* ... */]);
+            $this->send($entity->getRecipient(), 'Foo update', $body);
+        }
+    }
+
+    // src/AppBundle/Updates/BarUpdater.php
+    namespace AppBundle\Updates;
+
+    use AppBundle\BarManager;
+    use Psr\Container\ContainerInterface;
+
+    class BarUpdater extends AbstractUpdater
+    {
+        private $barManager;
+
+        public function __construct(ContainerInterface $locator, BarManager $barManager)
+        {
+            parent::__construct($locator);
+
+            $this->barManager = $barManager;
+        }
+
+        public function update($entity)
+        {
+            // ...
+
+            $body = $this->render('Emails/bar_update.html.twig', [/* ... */]);
+            $this->send($entity->getRecipient(), 'Bar update', $body);
+        }
+    }
+
+Optionally, you'll need to add the ``container.service_subscriber`` tag to
+the services definitions of the updaters to enable the service subscribers.
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        // app/config/services.yml
+        services:
+            AppBundle\Updates\:
+                resource: '../src/Updates'
+                tags: ['container.service_subscriber']
+
+        // or
+
+        services:
+            AppBundle\Updates\FooUpdater:
+                tags: ['container.service_subscriber']
+
+            AppBundle\Updates\BarUpdater:
+                tags: ['container.service_subscriber']
+                arguments:
+                    - '@Psr\Container\ContainerInterface'
+                    - '@AppBundle\BarManager'
+
+    .. code-block:: xml
+
+        <!-- app/config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+
+                <prototype namespace="AppBundle\Updates\" resource="../src/Updates">
+                    <tag name="container.service_subscriber" />
+                </prototype>
+
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        <code>
+
+Subscribed services
+-------------------
+
+Service subscribers rely on the ``getSubscribedServices()`` method to return a
+list of services types required to build a service locator for instances of
+that service subscriber::
+
+    use Psr\Log\LoggerInterface;
+
+    public static function getSubscribedServices()
+    {
+        return [
+            'AppBundle\FooInterface',
```

Comment:
Should be `FooInterface::class` with a `use` statement on top

Reply:
Was intended to show there are two ways to type out the FQCN but I can see this is unneccesary.